### PR TITLE
Feature/iterator

### PR DESCRIPTION
**Pull request by trasher**

### CODE REVIEW

File: inc/taskjobstate.class.php [hunk 1]
```diff
@@ -419,35 +419,41 @@ public function ajaxGetLogs($params)
...
+                    ]
+                ]
+            ],
+            'WHERE'  => [
+                'run.id' => $id,
+                'log.date' => ['<=' => $last_date]
+            ],
+            'ORDER'  => 'log.id DESC'
+        ]);
+
```
Comment (stonebuzz) [inline inc/taskjobstate.class.php:446]: Get some warning from ```deploy``` task (From computer tab)

image

Logs : 

```
[2023-05-04 09:58:29] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "logs" in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:58:29] glpiphplog.WARNING:   *** PHP Warning (2): Trying to access array offset on value of type null in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:58:29] glpiphplog.WARNING:   *** PHP Warning (2): Trying to access array offset on value of type null in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:58:29] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "logs" in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2059
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:58:29] glpiphplog.WARNING:   *** PHP Warning (2): foreach() argument must be of type array|object, null given in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 1521
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:59:38] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "logs" in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:59:38] glpiphplog.WARNING:   *** PHP Warning (2): Trying to access array offset on value of type null in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:59:38] glpiphplog.WARNING:   *** PHP Warning (2): Trying to access array offset on value of type null in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2058
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:59:38] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "logs" in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 2059
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1813 PluginGlpiinventoryDeployPackage->getMyDepoyPackagesState()
  .../glpiinventory/inc/deploypackage.class.php:1414 PluginGlpiinventoryDeployPackage->getPackageForMe()
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
[2023-05-04 09:59:38] glpiphplog.WARNING:   *** PHP Warning (2): foreach() argument must be of type array|object, null given in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/deploypackage.class.php at line 1521
  Backtrace :
  .../glpiinventory/inc/deploypackage.class.php:1097 PluginGlpiinventoryDeployPackage->showPackageForMe()
  src/CommonGLPI.php:689                             PluginGlpiinventoryDeployPackage::displayTabContentForItem()
  ajax/common.tabs.php:116                           CommonGLPI::displayStandardTab()
  public/index.php:82                                require()
  
```

File: inc/stateinventory.class.php [hunk 2]
```diff
@@ -116,16 +129,29 @@ public function display($options = [])
...
+            ],
+            'WHERE'  => [
+                'method' => 'networkinventory'
+            ],
+            'GROUPBY' => 'uniqid',
+            'ORDERBY' => 'uniqid DESC',
+            'LIMIT'  => (int)$_SESSION['glpilist_limit'],
+            'START'  => (int)$start
+        ]);
+
```
Comment (stonebuzz) [inline inc/stateinventory.class.php:153]: Some warning from this query  (GLPIInventory > Networking > Networking Inventory status)

image

```
[2023-05-04 10:04:06] glpiphplog.CRITICAL:   *** Uncaught Exception DivisionByZeroError: Division by zero in /home/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/glpiinventory/inc/stateinventory.class.php at line 236
  Backtrace :
  plugins/glpiinventory/front/stateinventory.php:50  PluginGlpiinventoryStateInventory->display()
  public/index.php:82                                require()
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]
I did a functional review with :
- NetDiscovery (Printer / Switch)
- NetInventory (Printer / Switch)

everything seems OK foreach

File: inc/taskjobstate.class.php (hunk 1 above)
Comment (trasher) [reply inc/taskjobstate.class.php]: Guess this is fixed as of c953f3d18924109f503e7128c1a9b4ae2dce1c78

File: inc/stateinventory.class.php (hunk 2 above)
Comment (trasher) [reply inc/stateinventory.class.php:151]: 3cae3eb2dc1eccd18831a7c64dd8bca0e5ed74c6 should fix the SQL issue, dunno if that will fix the division by 0 also

**Review by stonebuzz — APPROVED** [review]

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**